### ISSUE TITLE
[BUG] myloader causing metadata lock - table creation also stuck

### ISSUE THREAD

**Issue by madhurbhaiya** [mydumper/mydumper#993]
**Describe the bug**
I am using >=8 threads to load the data, and despite using --max-threads-per-table=1 argument, I can see a lot of threads executing simultaneously for single table. It has also locked out creation of a main table. And I can see lots of threads stuck on *Waiting for metadata lock*

**To Reproduce**
Command executed:

MyDumper command executed:

```
mydumper --host=source-ip --port=3306 --user=user -a --threads=16 --database=dbName --compress --compress-protocol --build-empty-files --routines --rows=50000 --verbose=3 --daemon --logfile=mydumper.log
```

MyLoader command executed:

```
myloader --host=127.0.0.1 --port=3306 --user=user --ask-password --threads=16 --max-threads-per-table=1 --max-threads-for-index-creation=1 --directory=dump-dir --database=user --queries-per-transaction=10000 --enable-binlog --verbose=3
```

What mydumper and myloader version has been used?

**mydumper0.13.1-2, built against MySQL 8.0.30-22 with SSL support with GZIP**

**myloader0.13.1-2, built against MySQL 8.0.30-22 with SSL support with GZIP**

I am using Percona Server for MySQL. Version is: **Ver 8.0.30-22 for Linux on x86_64 (Percona Server (GPL), Release '22', Revision '7e301439b65')**

**Expected behaviour**
1. Only a single process should have worked on a single table
2. Create table should have executed already before loading data in other tables. Especially, the table which is stuck for creation is a foreign key in most of the other tables.
3. Lot of threads are stuck in Waiting for metadata lock. This should not happen.

**Log**
```
** Message: 13:59:53.226: Thread 11: restoring `dbName`.`largeTable_1` part 6 of 16 from dbName.largeTable_1.00000.sql.gz. Progress 14 of 305. Using 16 of 1 threads.
** Message: 13:59:53.268: Thread 15: restoring `dbName`.`largeTable_1` part 1 of 16 from dbName.largeTable_1.00008.sql.gz. Progress 15 of 305. Using 16 of 1 threads.
** Message: 13:59:53.282: Thread 7: restoring `dbName`.`largeTable_1` part 16 of 16 from dbName.largeTable_1.00004.sql.gz. Progress 16 of 305. Using 16 of 1 threads.
** Message: 13:59:53.295: Thread 5: restoring `dbName`.`largeTable_1` part 11 of 16 from dbName.largeTable_1.00012.sql.gz. Progress 17 of 305. Using 16 of 1 threads.
** Message: 13:59:54.495: Thread 3: restoring `dbName`.`largeTable_1` part 14 of 16 from dbName.largeTable_1.00002.sql.gz. Progress 18 of 305. Using 16 of 1 threads.
** Message: 13:59:54.547: Thread 4: restoring `dbName`.`largeTable_1` part 10 of 16 from dbName.largeTable_1.00010.sql.gz. Progress 19 of 305. Using 16 of 1 threads.
** Message: 13:59:54.759: Thread 2: restoring `dbName`.`largeTable_1` part 2 of 16 from dbName.largeTable_1.00006.sql.gz. Progress 20 of 305. Using 16 of 1 threads.
** Message: 13:59:55.560: Thread 1: restoring `dbName`.`largeTable_1` part 12 of 16 from dbName.largeTable_1.00014.sql.gz. Progress 21 of 305. Using 16 of 1 threads.
** Message: 13:59:55.587: Thread 8: restoring `dbName`.`largeTable_1` part 4 of 16 from dbName.largeTable_1.00001.sql.gz. Progress 22 of 305. Using 16 of 1 threads.
** Message: 13:59:55.620: Thread 12: restoring `dbName`.`largeTable_1` part 7 of 16 from dbName.largeTable_1.00009.sql.gz. Progress 23 of 305. Using 16 of 1 threads.
** Message: 14:00:37.020: Thread 13: restoring `dbName`.`largeTable_1` part 8 of 16 from dbName.largeTable_1.00005.sql.gz. Progress 24 of 305. Using 16 of 1 threads.
** Message: 14:00:50.689: Thread 10: restoring `dbName`.`largeTable_1` part 5 of 16 from dbName.largeTable_1.00013.sql.gz. Progress 25 of 305. Using 16 of 1 threads.
** Message: 14:01:05.544: Thread 6: restoring `dbName`.`largeTable_1` part 9 of 16 from dbName.largeTable_1.00003.sql.gz. Progress 26 of 305. Using 16 of 1 threads.
** Message: 14:03:48.395: Thread 14: restoring `dbName`.`largeTable_1` part 13 of 16 from dbName.largeTable_1.00011.sql.gz. Progress 27 of 305. Using 16 of 1 threads.
```

**Environment (please complete the following information):**
 - OS version: Ubuntu 22.04.1 LTS (GNU/Linux 5.15.0-58-generic x86_64)
 - MyDumper version: mydumper0.13.1-2, built against MySQL 8.0.30-22 with SSL support with GZIP

**Additional context**
Result of **SHOW PROCESSLIST**

```
| 38779 | user    | 127.0.0.1:40206 | dbName         | Query            |      0 | freeing items                   | INSERT INTO `largeTable_1` (`amended_by`,`amended_in_2b`,`applicable_month`,`applicable_month |       511 |         0 |             0 |
| 38780 | user    | 127.0.0.1:40214 | dbName         | Query            |      0 | starting                        | NULL                                                                                                 |        28 |         0 |             0 |
| 38781 | user    | 127.0.0.1:40222 | dbName         | Query            |      0 | update                          | INSERT INTO `largeTable_1` (`amended_by`,`amended_in_2b`,`applicable_month`,`applicable_month |       313 |         0 |             0 |
| 38782 | user    | 127.0.0.1:40236 | dbName         | Query            |      0 | update                          | INSERT INTO `largeTable_1` (`amended_by`,`amended_in_2b`,`applicable_month`,`applicable_month |        91 |         0 |             0 |
| 38783 | user    | 127.0.0.1:40248 | dbName         | Query            |      0 | update                          | INSERT INTO `largeTable_1` (`amended_by`,`amended_in_2b`,`applicable_month`,`applicable_month |       199 |         0 |             0 |
| 38784 | user    | 127.0.0.1:40254 | dbName         | Query            |   1124 | Waiting for table metadata lock | INSERT INTO `largeTable_1` (`amended_by`,`amended_in_2b`,`applicable_month`,`applicable_month |   1124344 |         0 |             0 |
| 38785 | user    | 127.0.0.1:40260 | dbName         | Query            |      0 | update                          | INSERT INTO `largeTable_1` (`amended_by`,`amended_in_2b`,`applicable_month`,`applicable_month |       392 |         0 |             0 |
| 38786 | user    | 127.0.0.1:40274 | dbName         | Query            |      0 | update                          | INSERT INTO `largeTable_1` (`amended_by`,`amended_in_2b`,`applicable_month`,`applicable_month |       375 |         0 |             0 |
| 38787 | user    | 127.0.0.1:40288 | dbName         | Query            |      0 | update                          | INSERT INTO `largeTable_2` VALUES("10BDBPK3949K1ZN",2021,6,CONVERT("{\"data\": {\"lglnm\": \ |        29 |         0 |             0 |
| 38788 | user    | 127.0.0.1:40296 | dbName         | Query            |   1139 | Waiting for table metadata lock | INSERT INTO `largeTable_1` (`amended_by`,`amended_in_2b`,`applicable_month`,`applicable_month |   1139192 |         0 |             0 |
| 38789 | user    | 127.0.0.1:40302 | dbName         | Query            |      0 | update                          | INSERT INTO `largeTable_1` (`amended_by`,`amended_in_2b`,`applicable_month`,`applicable_month |       136 |         0 |             0 |
| 38790 | user    | 127.0.0.1:40312 | dbName         | Query            |   1194 | Waiting for table metadata lock | INSERT INTO `largeTable_1` (`amended_by`,`amended_in_2b`,`applicable_month`,`applicable_month |   1194268 |         0 |             0 |
| 38791 | user    | 127.0.0.1:40324 | dbName         | Query            |   1152 | Waiting for table metadata lock | INSERT INTO `largeTable_1` (`amended_by`,`amended_in_2b`,`applicable_month`,`applicable_month |   1152866 |         0 |             0 |
| 38792 | user    | 127.0.0.1:40330 | dbName         | Query            |    961 | Waiting for table metadata lock | INSERT INTO `largeTable_1` (`amended_by`,`amended_in_2b`,`applicable_month`,`applicable_month |    961490 |         0 |             0 |
| 38793 | user    | 127.0.0.1:40344 | dbName         | Query            |      0 | update                          | INSERT INTO `largeTable_1` (`amended_by`,`amended_in_2b`,`applicable_month`,`applicable_month |       380 |         0 |             0 |
| 38794 | user    | 127.0.0.1:40352 | dbName         | Query            |   1198 | Waiting for table metadata lock | CREATE TABLE `stuckTable_1` (
  `id` mediumint unsigned NOT NULL AUTO_INCREMENT,

```

**Comment by madhurbhaiya** [comment]
> Hhi @madhurbhaiya,
> 
> Why don't you try:
> 
> ```
> --serialized-table-creation          Table recreation will be executed in series, one thread at a time
> ```
> 
> ??

An update. 
Data loading eventually succeeded. The table stuck for creation eventually got created and loaded with data properly.

I will reattempt with --serialized-table-creation at later time.

**Comment by madhurbhaiya** [comment]
@davidducos --serialized-table-creation did help but not immediately

The "stuckTable_1" was still waiting for metadata lock for 15 seconds while other tables have been created and data restoring has started for them. However, it did get created much sooner this time.

I also changed one more thing this time. I removed these flags: ```--max-threads-per-table=1 --max-threads-for-index-creation=1```

However, I dont think these flags have any repercussions, because the very first time when I loaded the data, I was not using these max-threads-* flags and table creation was still stuck

For now I am closing the ticket. Thanks for the guidance.
